File: clients/csharp/src/Neo4j.AgentMemory/Models/Options.cs
using Neo4j.AgentMemory.Transport;

namespace Neo4j.AgentMemory.Models;

/// <summary>Transport selection mode.</summary>
public enum TransportMode
{
    /// <summary>Pick automatically based on endpoint shape (REST when /v1).</summary>
    Auto,
    /// <summary>TCK bridge protocol — POST /{snake_method}.</summary>
    Bridge,
    /// <summary>Hosted REST API at memory.neo4jlabs.com/v1.</summary>
    Rest,
}

/// <summary>Configuration options for creating a MemoryClient.</summary>
public class MemoryClientOptions
{
    /// <summary>URL of the memory service HTTP endpoint.</summary>
    public string Endpoint { get; set; } = "";

    /// <summary>Optional Bearer API key (e.g. nams_*).</summary>
    public string? ApiKey { get; set; }

    /// <summary>Optional token provider for OAuth refresh flows. Overrides ApiKey when set.</summary>
    public TokenProvider? TokenProvider { get; set; }

    /// <summary>Transport selection. Default: Auto.</summary>
    public TransportMode Transport { get; set; } = TransportMode.Auto;

    /// <summary>Additional headers added to every request.</summary>
    public IDictionary<string, string>? Headers { get; set; }

    /// <summary>Request timeout. Default: 30 seconds.</summary>
    public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(30);
}

File: clients/go/memory/options.go
package memory

import "time"

// Config holds the configuration for a memory Client.
type Config struct {
	// Endpoint is the base URL — bridge endpoint or `https://.../v1` for REST.
	Endpoint string

	// APIKey for Bearer authentication.
	APIKey string

	// TokenProvider returns a fresh access token. Overrides APIKey when set.
	TokenProvider TokenProvider

	// TransportMode picks between bridge / rest. Default is auto.
	TransportMode TransportMode

	// Headers added to every outgoing request.
	Headers map[string]string

	// Timeout for HTTP requests. Default: 30s.
	Timeout time.Duration

	// Namespace for multi-agent shared entity namespacing.
	Namespace string

	// CustomTransport overrides the auto-selected transport entirely.
	CustomTransport Transport
}

// Option configures a Client.
type Option func(*Config)

// WithEndpoint sets the HTTP endpoint URL.
func WithEndpoint(u string) Option {
	return func(c *Config) { c.Endpoint = u }
}

// WithAPIKey sets the authentication API key.
func WithAPIKey(key string) Option {
	return func(c *Config) { c.APIKey = key }
}

// WithTokenProvider supplies a callback that returns a fresh access token.
// Useful when running OAuth refresh-token flows.
func WithTokenProvider(p TokenProvider) Option {
	return func(c *Config) { c.TokenProvider = p }
}

// WithTimeout sets the HTTP request timeout.
func WithTimeout(d time.Duration) Option {
	return func(c *Config) { c.Timeout = d }
}

// WithNamespace sets the shared entity namespace.
func WithNamespace(ns string) Option {
	return func(c *Config) { c.Namespace = ns }
}

// WithTransportMode forces a specific transport.
func WithTransportMode(m TransportMode) Option {
	return func(c *Config) { c.TransportMode = m }
}

// WithHeaders adds custom HTTP headers.
func WithHeaders(h map[string]string) Option {
	return func(c *Config) { c.Headers = h }
}

// WithTransport supplies a custom Transport implementation.
func WithTransport(t Transport) Option {
	return func(c *Config) { c.CustomTransport = t }
}
